File: app/src/main/java/Square.java
import java.util.ArrayList;
import java.util.List;

public class Square {
    public static List<Square> allSquares = all();
    final int x;
    final int y;

    public Square(int x, int y) {
        this.x = x;
        this.y = y;
    }

    private static List<Square> all() {
        List<Square> squares = new ArrayList<>();

        for (int rank = 0; rank <= 7; rank++) {
            for (int file = 0; file <= 7; file++) {
                squares.add(new Square(rank, file));
            }
        }

        return squares;
    }

    Square nextRank(SetType set) {
        if (set == SetType.whiteSet) {
            return new Square(x - 1, y);
        } else {
            return new Square(x + 1, y);
        }
    }

    Square next2Rank(SetType set) {
        if (set == SetType.whiteSet) {
            return new Square(x - 2, y);
        } else {
            return new Square(x + 2, y);
        }
    }

    Square next2File(SetType set) {
        if (set == SetType.whiteSet) {
            return new Square(x, y + 2);
        } else {
            return new Square(x, y - 2);
        }
    }

    Square previous2File(SetType set) {
        if (set == SetType.whiteSet) {
            return new Square(x, y - 2);
        } else {
            return new Square(x, y + 2);
        }
    }

    Square nextRankNextFile(SetType set) {
        if (set == SetType.whiteSet) {
            return new Square(x - 1, y + 1);
        } else {
            return new Square(x + 1, y - 1);
        }
    }

    Square nextRankPreviousFile(SetType set) {
        if (set == SetType.whiteSet) {
            return new Square(x - 1, y - 1);
        } else {
            return new Square(x + 1, y + 1);
        }
    }

    Square previousFile(SetType set) {
        if (set == SetType.whiteSet) {
            return new Square(x, y - 1);
        } else {
            return new Square(x, y + 1);
        }
    }

    Square nextFile(SetType set) {
        if (set == SetType.whiteSet) {
            return new Square(x, y + 1);
        } else {
            return new Square(x, y - 1);
        }
    }

    public Square previousRank(SetType set) {
        if (set == SetType.whiteSet) {
            return new Square(x + 1, y);
        } else {
            return new Square(x - 1, y);
        }
    }

    public Square previous2Rank(SetType set) {
        if (set == SetType.whiteSet) {
            return new Square(x + 2, y);
        } else {
            return new Square(x - 2, y);
        }
    }

    public List<Square> otherSquaresInSameFile() {
        List<Square> acc = new ArrayList<>();
        for (int file = 0; file < 8; file++) {
            if (file != this.y) {
                acc.add(new Square(this.x, file));
            }
        }
        return acc;
    }

    public List<Square> otherSquaresInSameRank() {
        List<Square> acc = new ArrayList<>();
        for (int rank = 0; rank < 8; rank++) {
            if (rank != this.x) {
                acc.add(new Square(rank, this.y));
            }
        }
        return acc;
    }

    public boolean exists() {
        return this.x <= 7 && this.x >= 0 && this.y <= 7 && this.y >= 0;
    }

    public List<Square> diagonalSquares() {
        List<Square> acc = new ArrayList<>();
        for (int distance = 1; distance <= 7; distance++) {
            acc.add(new Square(x + distance, y + distance));
            acc.add(new Square(x + distance, y - distance));
            acc.add(new Square(x - distance, y + distance));
            acc.add(new Square(x - distance, y - distance));
        }
        return acc;
    }

    public List<Square> straightSquares() {
        List<Square> acc = new ArrayList<>();
        for (int distance = 1; distance <= 7; distance++) {
            acc.add(new Square(x + distance, y));
            acc.add(new Square(x - distance, y));
            acc.add(new Square(x, y + distance));
            acc.add(new Square(x, y - distance));
        }
        return acc;
    }
}

File: yagoc-desktop/src/main/kotlin/jenm/yagoc/board/Square.kt
package jenm.yagoc.board

import jenm.yagoc.pieces.PieceColor
import java.io.Serializable
import java.util.*
import java.util.stream.Collectors

data class Square(val arrayPosition: Int) : Serializable, Comparable<Square> {
    val rank = arrayPosition / 8
    val file = arrayPosition % 8

    fun nextRank(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.WhiteSet) {
            square(rank - 1, file)
        } else {
            square(rank + 1, file)
        }
    }

    fun next2Rank(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.WhiteSet) {
            square(rank - 2, file)
        } else {
            square(rank + 2, file)
        }
    }

    fun next2File(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.WhiteSet) {
            square(rank, file + 2)
        } else {
            square(rank, file - 2)
        }
    }

    fun previous2File(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.WhiteSet) {
            square(rank, file - 2)
        } else {
            square(rank, file + 2)
        }
    }

    fun nextRankNextFile(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.WhiteSet) {
            square(rank - 1, file + 1)
        } else {
            square(rank + 1, file - 1)
        }
    }

    fun nextRankPreviousFile(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.WhiteSet) {
            square(rank - 1, file - 1)
        } else {
            square(rank + 1, file + 1)
        }
    }

    fun previousFile(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.WhiteSet) {
            square(rank, file - 1)
        } else {
            square(rank, file + 1)
        }
    }

    fun nextFile(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.WhiteSet) {
            square(rank, file + 1)
        } else {
            square(rank, file - 1)
        }
    }

    fun previousRank(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.WhiteSet) {
            square(rank + 1, file)
        } else {
            square(rank - 1, file)
        }
    }

    fun previous2Rank(pieceColor: PieceColor): Square {
        return if (pieceColor == PieceColor.WhiteSet) {
            square(rank + 2, file)
        } else {
            square(rank - 2, file)
        }
    }

    fun exists(): Boolean {
        return rank in 0..7 && file in 0..7
    }

    fun diagonalSquares(): List<Square> {
        val acc: MutableList<Square> = ArrayList()
        for (distance in 1..7) {
            acc.add(square(rank + distance, file + distance))
            acc.add(square(rank + distance, file - distance))
            acc.add(square(rank - distance, file + distance))
            acc.add(square(rank - distance, file - distance))
        }
        return acc.stream().filter { obj: Square -> obj.exists() }.collect(Collectors.toList())
    }

    fun straightSquares(): List<Square> {
        val acc: MutableList<Square> = ArrayList()
        for (distance in 1..7) {
            acc.add(square(rank + distance, file))
            acc.add(square(rank - distance, file))
            acc.add(square(rank, file + distance))
            acc.add(square(rank, file - distance))
        }
        return acc.stream().filter { obj: Square -> obj.exists() }.collect(Collectors.toList())
    }

    override fun toString(): String {
        return "(" + Move.RANK_NAMES[rank] + ", " + Move.FILE_NAMES[file] + ')'
    }

    /**
     * position of the square in a sequence from 0 to 63
     */
    fun arrayPosition(): Int {
        return rank * 8 + file
    }

    override fun compareTo(other: Square): Int {
        return arrayPosition().compareTo(other.arrayPosition())
    }
}